### PR TITLE
Add support for signed hrefs

### PR DESCRIPTION
**Pull request by jmakovecki**
**Description:**
Adds support for modifying / signing links to Sentinel-3 products. Also includes some other minor fixes to docstrings, typing annotations, etc. 

**PR checklist:**

- [x] Code is formatted (run `scripts/format`).
- [x] Code lints properly (run `scripts/lint`).
- [x] Tests pass (run `scripts/test`).
- [x] Documentation has been updated to reflect changes, if applicable.
- [ ] Changes are added to the [CHANGELOG](../CHANGELOG.md).

### CODE REVIEW

File: src/stactools/sentinel3/file_extension_updated.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+              size: Optional[int] = None,
+              values: Optional[List[MappingObject]] = None,
+              local_path: Optional[str] = None) -> None:
+        super().apply(byte_order, checksum, header_size, size, values)
+        self.local_path = local_path
+
+    @property
+    def local_path(self) -> Optional[str]:
+        """Get or sets the local path of the file."""
+        return self._get_property(LOCAL_PATH_PROP, str)
```
Comment (gadomski) [inline src/stactools/sentinel3/file_extension_updated.py:25]: I'm not sure what the value of `local_path` is here. Most downstream users of these data will be hosting these STAC items over the network, so `local_path` would end up being pretty meaningless, I think?

**Review by gadomski — CHANGES_REQUESTED** [review]

File: src/stactools/sentinel3/file_extension_updated.py (hunk 1 above)
Comment (chorng) [reply src/stactools/sentinel3/file_extension_updated.py:25]: This is a property to reflect a relative path that can be used by clients for different purposes to organize the files locally, which is taken from [file info extension spec](https://github.com/stac-extensions/file#local-path).

Comment (gadomski) [reply src/stactools/sentinel3/file_extension_updated.py:25]: Can you give a use-case? I'm unclear how `local_path` would be used.

Comment (chorng) [reply src/stactools/sentinel3/file_extension_updated.py:25]: So, a relative path taken from product dir to individual files can be used If someone decides to copy the product files somewhere else, the local path would still point to the files from the project dir, whereas the full URL path would be invalid and the path would need to be parsed from it separately.

Another use case would be that [SNAP toolbox](https://step.esa.int/main/) reads the SAFE manifest to open a product that references the different files (e.g. image bands) relatively and thus expects to find them in the referenced folder (e.g. IMG_DATA).

Comment (gadomski) [reply src/stactools/sentinel3/file_extension_updated.py:25]: What is the `local_path` relative to? The item's self href? If so, how it it different than a relative `href`?

Your SAFE manifest example feels more like a "subasset" structure, where one asset references another asset -- but maybe I'm just not understanding the problem space correctly.

Do you have any open examples of tooling that uses `local_path`? That might help me understand better. I know **pystac** doesn't use it at all.

Comment (chorng) [reply src/stactools/sentinel3/file_extension_updated.py:25]: The stac items for Sentinel-3 on [Planetary Computer](https://planetarycomputer.microsoft.com/dataset/storage/sentinel-3) use this property. They have something like below:
```
"href" : "https://sentinel3euwest.blob.core.windows.net/sentinel-3/SLSTR/SL_2_FRP___/2023/03/10/S3B_SL_2_FRP____20230310T235936_20230311T000236_20230312T001332_0179_077_087_4500_PS2_O_NT_004.SEN3/cartesian_in.nc"
```
```
"file:local_path" : "S3B_SL_2_FRP____20230310T235936_20230311T000236_20230312T001332_0179_077_087_4500_PS2_O_NT_004.SEN3/cartesian_in.nc"
```

Comment (gadomski) [reply src/stactools/sentinel3/file_extension_updated.py:25]: Okay, after re-(and re-) reading the description in [the extension](https://github.com/stac-extensions/file#local-path) I guess I can see what it's used for. Feels pretty fragile to require files to be laid out a certain way on a filesystem but I guess that's how things are 🤷🏽.

@chorng thanks for your patience iterating this w/ me.

**Pull request merged**